flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/WindowUtil.scala

```diff
@@ -198,6 +204,79 @@ object WindowUtil {
     new TimeAttributeWindowingStrategy(windowSpec, timeAttributeType, timeIndex)
   }
 
+  /**
+   * Window TVF based aggregations don't support early-fire and late-fire,
+   * throws exception when the configurations are set.
+   */
+  def checkEmitConfiguration(tableConfig: TableConfig): Unit = {
+    val conf = tableConfig.getConfiguration
+    if (conf.getBoolean(TABLE_EXEC_EMIT_EARLY_FIRE_ENABLED) ||
+      conf.getBoolean(TABLE_EXEC_EMIT_LATE_FIRE_ENABLED)) {
+      throw new TableException("Currently, window table function based aggregate doesn't " +
+        s"support early-fire and late-fire configuration " +
+        s"'${TABLE_EXEC_EMIT_EARLY_FIRE_ENABLED.key()}' and " +
+        s"'${TABLE_EXEC_EMIT_LATE_FIRE_ENABLED.key()}'.")
+    }
+  }
+
+  // ------------------------------------------------------------------------------------------
+  // RelNode RowType
+  // ------------------------------------------------------------------------------------------
+
+  def deriveWindowAggregateRowType(
+      grouping: Array[Int],
+      aggCalls: Seq[AggregateCall],
+      windowing: WindowingStrategy,
+      namedWindowProperties: Seq[PlannerNamedWindowProperty],
+      inputRowType: RelDataType,
+      typeFactory: FlinkTypeFactory): RelDataType = {
+    val groupSet = ImmutableBitSet.of(grouping: _*)
+    val baseType = Aggregate.deriveRowType(
+      typeFactory,
+      inputRowType,
+      false,
+      groupSet,
+      Collections.singletonList(groupSet),
+      aggCalls)
+    val builder = typeFactory.builder
+    builder.addAll(baseType.getFieldList)
+    namedWindowProperties.foreach { namedProp =>
+      // use types from windowing strategy which keeps the precision and timestamp type
+      // cast the type to not null type, because window properties should never be null
+      val timeType = namedProp.getProperty match {
+        case _: PlannerWindowStart | _: PlannerWindowEnd =>
+          LogicalTypeUtils.removeTimeAttributes(windowing.getTimeAttributeType).copy(false)
+        case _: PlannerRowtimeAttribute | _: PlannerProctimeAttribute =>
+          windowing.getTimeAttributeType.copy(false)
+      }
+      builder.add(namedProp.getName, typeFactory.createFieldTypeFromLogicalType(timeType))
+    }
+    builder.build()
+  }
+
+  /**
+   * Derives output row type from local window aggregate
+   */
+  def deriveLocalWindowAggregateRowType(
+      aggInfoList: AggregateInfoList,
+      grouping: Array[Int],
+      inputRowType: RelDataType,
+      typeFactory: FlinkTypeFactory): RelDataType = {
+    val accTypes = aggInfoList.getAccTypes
+    val groupingTypes = grouping
+      .map(inputRowType.getFieldList.get(_).getType)
+      .map(FlinkTypeFactory.toLogicalType)
+    val sliceEndType = Array(DataTypes.BIGINT().getLogicalType)
+
+    val groupingNames = grouping.map(inputRowType.getFieldNames.get(_))
+    val accFieldNames = inferAggAccumulatorNames(aggInfoList)
+    val sliceEndName = Array("$slice_end")
```

Comment:
I'm not sure whether `slice_end` is proper field name now, because it could be window end or slice end.